### PR TITLE
Style Book: Fix `useForcedLayout` to prevent breaking style book

### LINKED ISSUE #8177: Style Book: Mini Cart Contents block displays a white box without content
## To reproduce

0. With WordPress 6.2 (or a previous version with GB installed) and a block theme like [Twenty Twenty-Three](https://wordpress.org/themes/twentytwentythree/) enabled.
1. Go to Appearance > Editor and edit a template.
2. In the toolbar, select Styles (black and white circle) and then, Open Style Book (eye icon).
3. Go to the WooCommerce tab.
4. Notice the Mini Cart Contents block displays a white box without content.

Mini Cart Contents block in Style Book

## Expected behavior

The block should display placeholder data so it can be previewed.

### LINKED ISSUE #8178: Style Book: Several filter blocks show an error
## To reproduce

0. With WordPress 6.2 (or a previous version with GB installed) and a block theme like [Twenty Twenty-Three](https://wordpress.org/themes/twentytwentythree/) enabled.
1. Go to Appearance > Editor and edit a template.
2. In the toolbar, select Styles (black and white circle) and then, Open Style Book (eye icon).
3. Go to the WooCommerce tab.
4. Notice the Filter by Price, Filter by Rating and Filter by Stock blocks show an error instead of the real block.

Filter by Price Controls block

## Expected behavior

The block should be displayed properly.

### LINKED ISSUE #8179: Style Book: Cart block shows an error: too much recursion
## To reproduce

0. With WordPress 6.2 (or a previous version with GB installed) and a block theme like [Twenty Twenty-Three](https://wordpress.org/themes/twentytwentythree/) enabled.
1. Go to Appearance > Editor and edit a template.
2. In the toolbar, select Styles (black and white circle) and then, Open Style Book (eye icon).
3. Go to the WooCommerce tab.
4. Notice the Cart block displays an error.

Cart block in Style Book

## Expected behavior

The Cart block should be displayed correctly without errors.

### PR DESCRIPTION
**Pull request by opr**
In this PR we will update `useForcedLayout` to check if it has already dispatched `replaceInnerBlocks` - if it has we won't do that again.

The hook subscribes to the block editor registry and the callback is executed when a data store changes. We updated the second argument of `subscribe` so it only listens for changes on `core/block-editor`.

When the callback is executed, if the current block does not have any inner blocks, but it should, then `replaceInnerBlocks` is dispatched, causing these blocks to be added. Once they are added, then this shouldn't need to be done again, that is why we introduced the `templatesSynced` variable to track this, per hook.

Fixes #8179
Fixes #8177
Fixes #8178

#### Internal developer testing - do not include these in testing notes.

1. In your `wp-content/plugins` directory, run: `npx @wordpress/create-block -t @woocommerce/extend-cart-checkout-block newsletter-plugin`
2. Go to WP -> Plugins and activate the "Newsletter Plugin"
3. Add a new page, and add the Checkout block.
4. Ensure that in the contact information section, a checkbox for subscribing to the newsletter is added.
5. Save the page, add something to your cart, then open the page in the front-end.
6. Check the box is also there on the front-end.

#### User Facing Testing

1. With Gutenberg installed and a block theme like [Twenty Twenty-Three](https://wordpress.org/themes/twentytwentythree/) enabled.
2. Go to Appearance > Editor and edit a template.
3. In the toolbar, select Styles (black and white circle) and then, Open Style Book (eye icon).
4. Go to the WooCommerce tab.
5. Ensure the Mini Cart block load correctly.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

This should have a slight performance increase in the editor due to `useForcedLayout` doing less work once templates have already been synced.

### Changelog

> Fix: Resolve an issue where the WooCommerce tab of the style book would crash and certain blocks would not load correctly.

### CODE REVIEW

**Review by Aljullu — APPROVED** [review]
Thanks for working on this, @opr!

Confirming it fixes #8179. It also fixes #8177 and #8178, I updated the PR description accordingly.

Code changes LGTM as well, but I would appreciate a second look from somebody from Rubik with more expertise in this part of the codebase.

File: assets/js/blocks/cart-checkout-shared/use-forced-layout.ts [hunk 1]
```diff
@@ -116,6 +115,7 @@ export const useForcedLayout = ( {
 
 	const registry = useRegistry();
 	useEffect( () => {
+		let templateSynced = false;
```
Comment (senadir) [inline assets/js/blocks/cart-checkout-shared/use-forced-layout.ts:118]: Wouldn't `templateSynced` get reset on every re-render, or is the `useRegistry` here stable?

**Review by senadir — APPROVED** [review]
Logic sounds fine to me.

File: assets/js/blocks/cart-checkout-shared/use-forced-layout.ts (hunk 1 above)
Comment (opr) [reply assets/js/blocks/cart-checkout-shared/use-forced-layout.ts:118]: It seems stable, logging this happens only once when a block is rendered for the first time.

**Pull request merged**